libs/agno/agno/agent/agent.py

```diff
@@ -2022,6 +2022,28 @@ async def _amake_memories_and_summaries(
             log_debug("Creating session summary.")
             await self.memory.acreate_session_summary(session_id=session_id, user_id=user_id)
 
+    def _raise_if_async_tools(self) -> None:
+        """Raise an exception if any tools contain async functions"""
+        if self.tools is None:
+            return
+
+        from inspect import iscoroutinefunction
+
+        for tool in self.tools:
+            if isinstance(tool, Toolkit):
+                for func in tool.functions:
+                    if iscoroutinefunction(tool.functions[func].entrypoint):
+                        raise Exception(
+                            f"Async tool {tool.name} can't be used with synchronous agent.run() or agent.print_response(). "
+                            "Use agent.arun() or agent.aprint_response() instead to use this tool."
+                        )
+            elif isinstance(tool, Callable) or isinstance(tool, Function):
+                if iscoroutinefunction(tool):
+                    raise Exception(
+                        f"Async function {tool.__name__} can't be used with synchronous agent.run() or agent.print_response(). "
+                        "Use agent.arun() or agent.aprint_response() instead to use this tool."
+                    )
```

Comment:
@dirkbrnd 
1. Looks like self.tools can also be dict. How can I trigger such case? I'll update this logic when I figure that out
2. Raising here sounds like better dx (user will see the log plus we avoid any further disfunction), but lmk if we prefer to warn and remove the async tool.

Reply:
1. That is in cases for native tools that we have on models. We can't check that, so its ok to skip.

2. Yeah lets raise. There shouldn't be a legitimate reason for the user to continue.